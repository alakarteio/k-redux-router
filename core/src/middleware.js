/* eslint-env browser */

// https://stackoverflow.com/questions/1714786/query-string-encoding-of-a-javascript-object
const toQueryString = obj => Object.keys(obj).map(k => `${encodeURIComponent(k)}=${encodeURIComponent(obj[k])}`).join('&')

// https://stackoverflow.com/questions/8648892/convert-url-parameters-to-a-javascript-object
const queryToObject = () => {
  const { location } = window

  if (location.search.length < 2) return {}

  const search = location.search.substring(1)
  return JSON.parse(`{"${decodeURI(search).replace(/"/g, '\\"').replace(/&/g, '","').replace(/=/g, '":"')}"}`)
}

const routeFound = result => ({ type: '@@router/ROUTE_FOUND', payload: Object.assign({ found: true }, result) })

const dispatchResultFactory = reducer => (store) => {
  if (window && window.location && window.location.pathname) {
    // find route (& path params)
    let pathParams
    const route = reducer.getState(store.getState()).routes.array.find((r) => {
      pathParams = r.href.regexp.exec(window.location.pathname)
      return !!pathParams
    })

    // attach names to path params
    pathParams = route.href.parsed
      .map(part => part.name)
      .filter(Boolean)
      .reduce(
        (acc, curr, index) => Object.assign({ [curr]: pathParams[index + 1] }, acc),
        {},
      )

    store.dispatch(routeFound({
      route,
      params: {
        path: pathParams,
        query: queryToObject(),
      },
    }))
  }
}

const mapActionFactory = reducer => store => (action) => {
  const { type, payload } = action

  switch (type) {
    case '@@router/GO_BACK': {
      window.history.go(payload * -1)
      return undefined
    }
    case '@@router/REPLACE':
    case '@@router/PUSH': {
      const { code, params } = payload

      // find route
      const route = reducer.getState(store.getState()).routes.map[code]

      // update history API
      if (route) {
        const { href } = route
        let queryPart = ''

        let toPush = href.base
        if (href.compiled) toPush = href.compiled(params.path)
        if (params.query) queryPart = `?${toQueryString(params.query)}`

        toPush = `${toPush}${queryPart}`
        if (type === '@@router/PUSH') window.history.pushState(undefined, undefined, toPush)
        else window.history.replaceState(undefined, undefined, toPush)
      }

      // TODO: if no route, find the closest `notFound` to push it in `result`

      // update state
      return routeFound({
        route,
        params,
      })
    }
    // default is undefined so we avoid infinite loop
    default: return undefined
  }
}

export default (routes, options, reducer) => {
  let initialized = false
  const dispatchResult = dispatchResultFactory(reducer)
  const mapAction = mapActionFactory(reducer)

  // redux middleware
  return store => next => (action) => {
    // these action are not catched by the middleware and are used as it is
    if (!action.type || !action.type.startsWith('@@router/')) return next(action)

    // dispatch base action so other lib and reducer can plug to it
    const res = next(action)

    // init watchers (location/INIT)
    // - this is done here because we can't dispatch when middlewares tree is initialized by redux
    // - so we make sure that the dispatch are done when we are initialized
    if (action.type === '@@router/INIT') {
      dispatchResult(store) // TODO: move to store.dispatch pattern

      if (initialized) {
        console.warn('[k-redux-router] initialized twice') // eslint-disable-line no-console
      } else {
        initialized = true

        // watcher (location)
        window.onpopstate = () => { dispatchResult(store) } // TODO: move to store.dispatch pattern
      }
    }

    if (!initialized && action.type !== '@@router/ROUTE_FOUND') console.warn('[k-redux-router] router should be initialized') // eslint-disable-line no-console

    // router actions can found a new route
    const newAction = mapAction(store)(action)
    if (newAction) store.dispatch(newAction)

    return res
  }
}
