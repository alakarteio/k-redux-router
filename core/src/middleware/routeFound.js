const type = '@@router/ROUTE_FOUND'

export default Object.assign(
  result => ({ type, payload: Object.assign({ found: true }, result) }),
  {
    type,
  },
)
