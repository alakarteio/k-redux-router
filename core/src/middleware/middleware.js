/* eslint-env browser */
import dispatchResultFactory from './dispatchResult'
import mapActionFactory from './mapAction'
import routeFound from './routeFound'

export default (routes, options, reducer) => {
  let initialized = false
  const dispatchResult = dispatchResultFactory(options, reducer)
  const mapAction = mapActionFactory(options, reducer)

  // redux middleware
  return store => next => (action) => {
    // these action are not catched by the middleware and are used as it is
    if (!action.type || !action.type.startsWith('@@router/')) return next(action)

    // dispatch base action so other lib and reducer can plug to it
    const res = next(action)

    // init watchers (location/INIT)
    // - this is done here because we can't dispatch when middlewares tree is initialized by redux
    // - so we make sure that the dispatch are done when we are initialized
    if (action.type === '@@router/INIT') { // TODO: do like routeFound.type
      dispatchResult(store) // TODO: move to store.dispatch pattern

      if (initialized) {
        console.warn('[k-redux-router] initialized twice') // eslint-disable-line no-console
      } else {
        initialized = true

        // watcher (location)
        // TODO: make it pure
        if (window) window.onpopstate = () => { dispatchResult(store) } // TODO: move to store.dispatch pattern
      }
    }

    if (!initialized && action.type !== routeFound.type) console.warn('[k-redux-router] router should be initialized') // eslint-disable-line no-console

    // router actions can found a new route
    const newAction = mapAction(store)(action)
    if (newAction) store.dispatch(newAction)

    return res
  }
}
