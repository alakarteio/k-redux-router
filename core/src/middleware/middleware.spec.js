/* eslint-disable import/first */
/* eslint-env jest */
jest.mock('./dispatchResult')
jest.mock('./routeFound')

import middleware from './middleware'
import dispatchResultFactory from './dispatchResult'

describe('middleware', () => {
  const reducer = {
    getState: state => state.router,
  }

  const options = {
    history: {
      pushState: jest.fn(),
    },
  }

  const state = {
    dummy: { state: 'true' },
    router: {
      routes: {
        map: {
          code: {
            code: 'route-code',
            isRoute: true,
            href: {
              base: '/base/path',
            },
          },
        },
      },
    },
  }
  const store = {
    store: true,
    getState: () => state,
    dispatch: jest.fn(),
  }

  const next = jest.fn(() => 'next result')

  const routes = {
    '/': {
      code: 'main',
      '/users': {
        code: 'users',
        '/:id': {
          code: 'user',
        },
      },
    },
  }

  beforeEach(() => {
    next.mockClear()
    store.dispatch.mockClear()
    options.history.pushState.mockClear()
  })

  const match = action => () => {
    const mockedDispatchResult = jest.fn()
    dispatchResultFactory.mockImplementation(() => mockedDispatchResult)

    const res = middleware(routes, options, reducer)(store)(next)(action)

    expect({
      res,
      next: next.mock.calls,
      dispatch: store.dispatch.mock.calls,
      dispatchResult: mockedDispatchResult.mock.calls,
      pushState: options.history.pushState.mock.calls,
    }).toMatchSnapshot()
  }

  it('should do nothing but call next middleware -action without type-', match({ action: 'without type' }))
  it('should do nothing but call next middlware -action that is not router scope-', match({ type: 'NOT-ROUTER' }))
  it('should init the router', match({ type: '@@router/INIT' }))
  it('should dispatch router action', match({ type: '@@router/PUSH', payload: { code: 'code' } }))
})
