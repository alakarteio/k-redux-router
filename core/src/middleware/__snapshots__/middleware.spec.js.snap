// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware should dispatch router action 1`] = `
Object {
  "dispatch": Array [],
  "dispatchResult": Array [],
  "next": Array [
    Array [
      Object {
        "payload": Object {
          "code": "code",
        },
        "type": "@@router/PUSH",
      },
    ],
  ],
  "pushState": Array [
    Array [
      undefined,
      undefined,
      "/base/path",
    ],
  ],
  "res": "next result",
}
`;

exports[`middleware should do nothing but call next middleware -action without type- 1`] = `
Object {
  "dispatch": Array [],
  "dispatchResult": Array [],
  "next": Array [
    Array [
      Object {
        "action": "without type",
      },
    ],
  ],
  "pushState": Array [],
  "res": "next result",
}
`;

exports[`middleware should do nothing but call next middlware -action that is not router scope- 1`] = `
Object {
  "dispatch": Array [],
  "dispatchResult": Array [],
  "next": Array [
    Array [
      Object {
        "type": "NOT-ROUTER",
      },
    ],
  ],
  "pushState": Array [],
  "res": "next result",
}
`;

exports[`middleware should init the router 1`] = `
Object {
  "dispatch": Array [],
  "dispatchResult": Array [
    Array [
      Object {
        "dispatch": [MockFunction],
        "getState": [Function],
        "store": true,
      },
    ],
  ],
  "next": Array [
    Array [
      Object {
        "type": "@@router/INIT",
      },
    ],
  ],
  "pushState": Array [],
  "res": "next result",
}
`;
